### PR TITLE
Add ability to mount an existing filesystem 

### PR DESCRIPTION
**Pull request by junjihashimoto**
Hi,

This PR is adding new Filesystem(PROXYFS) which can mount an existing filesystem .
I read https://github.com/kripken/emscripten/issues/2963.

Usage is below.
FS mounts other FS module.

```
var m = require('./something-child-module');
FS.mkdir('/working');
FS.mount(PROXYFS, { root: '/' , fs:m.FS}, '/working');
```

What do you think?

### CODE REVIEW

**Comment by kripken** [comment]
This is interesting, but I don't quite see the benefit yet. What is an example use case that could benefit from this?

**Comment by junjihashimoto** [comment]
I am thinking multiple programs which have own main-function work together.
These programs share same files.

In my use case, these programs are vim(vimjs), hugs(haskell interpreter) and gnuplot.
The programs(vim and hugs) have own assets(/usr/share/vim or /usr/share/hugs).
vim writes source code, then hugs runs the code.
And vim reads some results which hugs outputs.

I tried to think following cases to share files.

Case1, Assign FS to global variable.
*  Method: Program A and Program use the same file system.
*  Pros: Small resources(no duplicated data)
*  Cons: Current emscripten does not support, so we have to modifiy generated js-program directly.
        Global FS may make some trouble (ex. symbol conflict).

Case2, Use BrowserFS
*  Method: Mount dropbox-fs of BrowserFS which can share files among programs.
*  Pros: No use global FS.
        We just mount BrowserFS.
*  Cons: Each program can not access other program assets.
        BrowserFS can mount new file-tree which may be empty at first.
        (In my case, vim can not change hugs's settings)
        dropbox is not general method, why not google-drive or some others.

Case3, Use ProxyFS
* Method: Mount FS on other program.
* Pros: No use global FS.
        We just mount other FS.
        We do not need other FS except for ProxyFS.
* Cons: Each program can access other program assets.
        (In my case, vim can change hugs's settings)

I choose case3 and made ProxyFS.

ProxyFS's benefit is below.

1, Multi programs can share same files.
    A program can access assets of other program.
2, We do not need to change mounted program and its file-system.
    So modularity is good.
   For example, programs-A-FS mounts programs-B-FS. programs-B-FS is not changed.
3, We do not need external file-system like browserFS to share files except for ProxyFS.

**Comment by kripken** [comment]
About (1), I think that should work. If FS is exported, you can get Module['FS'] from one, and assign it in the other (using a pre-js or EM_ASM, as the code assigning to FS needs to have FS in its scope). Unless I am forgetting some problem?

About (2), can't BrowserFS let you share files with all its backends, not just dropbox-fs?

**Comment by junjihashimoto** [comment]
About (1), I have tried shared file by pre-js([the code](https://gist.github.com/junjihashimoto/a44d7067e3d326fe739f9c79f8d16a89)), but it does not work.
Then I tried shared file by global-variable([the code](https://gist.github.com/junjihashimoto/a88b6f1c543a612583561ab99bfd7824)). This works.
I suspect  FS variable is locally bound.

In addition to this, file descriptor of a program is bound to FS, because file descriptor is in FS module like [this](https://github.com/kripken/emscripten/blob/master/src/library_fs.js#L377).
I think that If there are multiple main functions, the programs(functions) may cause a problem.

About (2), I agree with you. BrowerFS is not file system to share files. The file system may make some problems. (Actually, I do not try BrowerFS.)

So, case(3) is best solution to share files. Don't you think so?

**Comment by junjihashimoto** [comment]
About (1), let me explain it is something wrong that  file descriptor of a program is bound to FS.
When program A and B are connected by pipe( for example, these commands are grep | sort " ).
If we write something to file descriptor=0(normally stdin), both program A and B recieve it at the same time, because program A and B share file descriptor, even if we want to send a data to one side of pipe at first.
So when there are multi programs and these are connected by pipe, file descriptor space of programs must be splited each other.

**Comment by kripken** [comment]
I see, thank you. Yes, that makes sense, I understand the approach now.

File: tests/test_core.py [hunk 1]
```diff
@@ -7179,6 +7179,235 @@ def test_mallinfo(self):
...
+  }
+  return 0;
+}
+''')
+
+    Popen([PYTHON, EMCC,
+           '-o', 'proxyfs_test.js', 'proxyfs_test.c',
+           '--embed-file', 'proxyfs_embed.txt', '--pre-js', 'proxyfs_pre.js',
+           '-s', 'MAIN_MODULE=1']).communicate()
+    Popen(['cp', 'proxyfs_test.js', 'proxyfs_test1.js']).communicate()
```
Comment (kripken) [inline tests/test_core.py]: `cp` won't work on windows. you can use `shutil.copyfile` instead.

File: tests/test_core.py [hunk 2]
```diff
@@ -7179,6 +7179,235 @@ def test_mallinfo(self):
   def test_wrap_malloc(self):
     self.do_run(open(path_from_root('tests', 'wrap_malloc.cpp')).read(), 'OK.')
 
+
+  def test_proxyfs(self):
```
Comment (kripken) [inline tests/test_core.py]: the test should be in `test_other.py` as it does not need to run in all different optimization modes

File: tests/test_core.py [hunk 3]
```diff
@@ -7179,6 +7179,235 @@ def test_mallinfo(self):
...
+  return 0;
+}
+''')
+
+    Popen([PYTHON, EMCC,
+           '-o', 'proxyfs_test.js', 'proxyfs_test.c',
+           '--embed-file', 'proxyfs_embed.txt', '--pre-js', 'proxyfs_pre.js',
+           '-s', 'MAIN_MODULE=1']).communicate()
+    Popen(['cp', 'proxyfs_test.js', 'proxyfs_test1.js']).communicate()
+    Popen(['cp', 'proxyfs_test.js', 'proxyfs_test2.js']).communicate()
```
Comment (kripken) [inline tests/test_core.py]: i feel like i'm missing something here. if the js files are all just copies of each other, how do they do different things?

**Comment by junjihashimoto** [comment]
Thank you for your reviewing my codes.
Updating test is done.
- ```cp``` is replaced with ````shutil.copyfile```.
-  test codes is moved to test_other.py. 
-  Add a reason for just copying proxyfs_test.js to proxyfs_test[12].js. 
    This  prevents 'require' of node.js from caching js-object. 
　In this test, I suppose that 3 different programs share the same directory and files.
   When caching is available, a object of  first ```require('js-files')``` is the same as one of second ```require('js-files')```.

File: src/library_proxyfs.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+          if (!e.code) throw e;
+          throw new FS.ErrnoError(ERRNO_CODES[e.code]);
+        }
+        return node;
+      },
+      rename: function (oldNode, newDir, newName) {
+        var oldPath = PROXYFS.realPath(oldNode);
+        var newPath = PATH.join2(PROXYFS.realPath(newDir), newName);
+        try {
+          oldNode.moount.opts.fs.rename(oldPath, newPath);
```
Comment (kripken) [inline src/library_proxyfs.js]: looks like a typo here, `moount`

File: src/library_proxyfs.js [hunk 5]
```diff
@@ -100,7 +100,7 @@ mergeInto(LibraryManager.library, {
         var oldPath = PROXYFS.realPath(oldNode);
         var newPath = PATH.join2(PROXYFS.realPath(newDir), newName);
         try {
-          oldNode.moount.opts.fs.rename(oldPath, newPath);
+          oldNode.mount.opts.fs.rename(oldPath, newPath);
```
Comment (junjihashimoto) [inline src/library_proxyfs.js]: Fix typo
Thank you for finding this.

File: tests/test_other.py [hunk 6]
```diff
@@ -2599,6 +2599,22 @@ def test_fs_stream_proto(self):
...
+    #   var m0 = require('./proxyfs_test.js');
+    #   var m1 = require('./proxyfs_test.js');
+    #   var m2 = require('./proxyfs_test.js');
+    #
+    # To separate js-objects for each of them, following 'require' use different js-files.
+    #
+    #   var m0 = require('./proxyfs_test.js');
+    #   var m1 = require('./proxyfs_test1.js');
+    #   var m2 = require('./proxyfs_test2.js');
+    #
```
Comment (junjihashimoto) [inline tests/test_other.py]: Add reason of loading multiple code.

**Comment by junjihashimoto** [comment]
Rebase is done.(It is because src/module.js of emscripten-strict conflicted. )

**Comment by kripken** [comment]
Looks good, thank you. Ok, this is ready to land, but we need to wait a little because `incoming` is frozen until we merge it to master. Hopefully that'll happen very soon though, and we'll merge this and other pending PRs.

**Comment by junjihashimoto** [comment]
Thank you too.
Best wishes to you for a happy Christmas and a joyful New Year.

**Pull request merged**